dm/dm/worker/source_worker_test.go

```diff
@@ -759,3 +759,72 @@ func (t *testServer) testSourceWorker(c *C) {
 	err = w.OperateSubTask("testSubTask", pb.TaskOp_Delete)
 	c.Assert(err, ErrorMatches, ".*worker already closed.*")
 }
+
+func (t *testServer) TestQueryValidator(c *C) {
+	cfg := loadSourceConfigWithoutPassword(c)
+
+	dir := c.MkDir()
+	cfg.EnableRelay = true
+	cfg.RelayDir = dir
+	cfg.MetaDir = dir
+
+	var (
+		masterAddr   = tempurl.Alloc()[len("http://"):]
+		keepAliveTTL = int64(1)
+	)
+	etcdDir := c.MkDir()
+	ETCD, err := createMockETCD(etcdDir, "http://"+masterAddr)
+	c.Assert(err, IsNil)
+	defer ETCD.Close()
+	workerCfg := NewConfig()
+	c.Assert(workerCfg.Parse([]string{"-config=./dm-worker.toml"}), IsNil)
+	workerCfg.Join = masterAddr
+	workerCfg.KeepAliveTTL = keepAliveTTL
+	workerCfg.RelayKeepAliveTTL = keepAliveTTL
+
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:            GetJoinURLs(workerCfg.Join),
+		DialTimeout:          dialTimeout,
+		DialKeepAliveTime:    keepaliveTime,
+		DialKeepAliveTimeout: keepaliveTimeout,
+	})
+	c.Assert(err, IsNil)
+
+	w, err := NewSourceWorker(cfg, etcdCli, "", "")
+	w.closed.Store(false)
+	c.Assert(err, IsNil)
+	c.Assert(failpoint.Enable("github.com/pingcap/tiflow/dm/syncer/MockValidationQuery", `return(true)`), IsNil)
+	c.Assert(failpoint.Enable("github.com/pingcap/tiflow/dm/dm/worker/WorkerMockValidationQuery", `return(true)`), IsNil)
+	c.Assert(w.StartSubTask(&config.SubTaskConfig{
```

Comment:
maybe we can `recordSubTask` in subtask holder, so we don't need to start the task directly. so we can remove the hacky failpoint in source worker.